OpenRA.Mods.Common/Widgets/Logic/MultiplayerLogic.cs

```diff
@@ -356,6 +375,85 @@ void SelectServer(GameServer server)
 		{
 			currentServer = server;
 			currentMap = server != null ? modData.MapCache[server.Map] : null;
+
+			clientList.RemoveChildren();
+			if (server == null || !server.Clients.Any())
+			{
+				joinButton.Bounds.Y = joinButtonY;
+				return;
+			}
+
+			joinButton.Bounds.Y = clientContainer.Bounds.Bottom;
+
+			var players = server.Clients
+				.Where(c => !c.IsSpectator)
+				.GroupBy(p => p.Team)
+				.OrderBy(g => g.Key);
+
+			var teams = new Dictionary<string, IEnumerable<GameClient>>();
+			var noTeams = players.Count() == 1;
+			foreach (var p in players)
+			{
+				var label = noTeams ? "Players" : p.Key == 0 ? "No Team" : "Team {0}".F(p.Key);
+				teams.Add(label, p);
+			}
+
+			if (server.Clients.Any(c => c.IsSpectator))
+				teams.Add("Spectators", server.Clients.Where(c => c.IsSpectator));
+
+			// Can only show factions if the server is running the same mod
+			var disableFactionDisplay = server.Mod != modData.Manifest.Id;
+
+			if (mapPreview != null)
+			{
+				var spawns = currentMap.SpawnPoints;
+				var occupants = server.Clients
+					.Where(c => (c.SpawnPoint - 1 >= 0) && (c.SpawnPoint - 1 < spawns.Length))
+					.ToDictionary(c => spawns[c.SpawnPoint - 1], c => new SpawnOccupant(c, disableFactionDisplay));
+
+				mapPreview.SpawnOccupants = () => occupants;
+			}
+
+			var factionInfo = modData.DefaultRules.Actors["world"].TraitInfos<FactionInfo>();
+			foreach (var kv in teams)
+			{
+				var group = kv.Key;
+				if (group.Length > 0)
+				{
+					var header = ScrollItemWidget.Setup(clientHeader, () => true, () => { });
+					header.Get<LabelWidget>("LABEL").GetText = () => group;
+					clientList.AddChild(header);
+				}
+
+				foreach (var option in kv.Value)
+				{
+					var o = option;
+
+					var item = ScrollItemWidget.Setup(clientTemplate, () => false, () => { });
+					if (!o.IsSpectator && !disableFactionDisplay)
+					{
+						var label = item.Get<LabelWidget>("LABEL");
+						var font = Game.Renderer.Fonts[label.Font];
+						var name = WidgetUtils.TruncateText(o.Name, label.Bounds.Width, font);
+						label.GetText = () => name;
+						label.GetColor = () => o.Color.RGB;
+
+						var flag = item.Get<ImageWidget>("FLAG");
+						flag.IsVisible = () => true;
+						flag.GetImageCollection = () => "flags";
+						flag.GetImageName = () => (factionInfo != null && factionInfo.Any(f => f.InternalName == o.Faction)) ? o.Faction : "Random";
+					}
+					else
+					{
+						var label = item.Get<LabelWidget>("NOFLAG_LABEL");
+						var font = Game.Renderer.Fonts[label.Font];
+						var name = WidgetUtils.TruncateText(o.Name, label.Bounds.Width, font);
+						label.GetText = () => name;
```

Comment:
Any reason to not color the player name here as well?

Reply:
Done.